library/psa_crypto.c

```diff
@@ -3131,7 +3156,761 @@ psa_status_t psa_asymmetric_decrypt( mbedtls_svc_key_id_t key,
+    operation->num_ops = 0;
+
+    status = psa_driver_wrapper_verify_hash_start( operation, &attributes,
+                                                  slot->key.data,
+                                                  slot->key.bytes,
+                                                  alg, hash, hash_length,
+                                                  signature, signature_length );
+
+    if( status != PSA_SUCCESS )
+    {
+        operation->error_occurred = 1;
+        psa_verify_hash_abort_internal( operation );
+    }
+
+    unlock_status = psa_unlock_key_slot( slot );
+
+    if( unlock_status != PSA_SUCCESS )
+        operation->error_occurred = 1;
+
+    return( ( status == PSA_SUCCESS ) ? unlock_status : status );
+}
+
+psa_status_t psa_verify_hash_complete(
+        psa_verify_hash_interruptible_operation_t *operation)
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    /* Check that start has been called first, and that operation has not
+     * previously errored. */
+    if( operation->id == 0 || operation->error_occurred )
+    {
+        status = PSA_ERROR_BAD_STATE;
+        goto exit;
+    }
+
+    status = psa_driver_wrapper_verify_hash_complete(operation);
+
+    /* Update ops count with work done. */
+    operation->num_ops = psa_driver_wrapper_verify_hash_get_num_ops(
+            operation);
 
+exit:
+
+    if( status != PSA_OPERATION_INCOMPLETE )
+    {
+        if( status != PSA_SUCCESS )
+            operation->error_occurred = 1;
+
+        psa_verify_hash_abort_internal(operation);
+    }
+
+    return( status );
+}
+
+psa_status_t psa_verify_hash_abort(
+                        psa_verify_hash_interruptible_operation_t *operation )
+{
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    status = psa_verify_hash_abort_internal( operation );
+
+    /* We clear the number of ops done here, so that it is not cleared when
+     * the operation fails or succeeds, only on manual abort. */
+    operation->num_ops = 0;
+
+    /* Likewise, failure state. */
+    operation->error_occurred = 0;
+
+    return( status );
+}
+
+/****************************************************************/
+/* Asymmetric interruptible cryptography internal               */
+/* implementations                                              */
+/****************************************************************/
+
+static uint32_t mbedtls_psa_interruptible_max_ops =
+                                            PSA_INTERRUPTIBLE_MAX_OPS_UNLIMITED;
+
+void mbedtls_psa_interruptible_set_max_ops( uint32_t max_ops )
+{
+    mbedtls_psa_interruptible_max_ops = max_ops;
+
+#if (defined(MBEDTLS_PSA_BUILTIN_ALG_ECDSA) || \
+     defined(MBEDTLS_PSA_BUILTIN_ALG_DETERMINISTIC_ECDSA) ) && \
+     defined( MBEDTLS_ECP_RESTARTABLE )
+
+    /* Internal implementation uses zero to indicate infinite number max ops,
+     * therefore avoid this value, and set to minimum possible. */
+    if( max_ops == 0 )
+        max_ops = 1;
```

Comment:
[non-blocker] This is not tested according to lcov, but it seems testable.

Reply:
#7113